### PR TITLE
feat(guardrails): add NeMo Provider

### PR DESCRIPTION
**Pull request by christinaexyou**
Refactor of https://github.com/praxis-proxy/praxis/pull/700

Implements the NeMo provider for the ai_guardrails filter, completing request-side evaluation against NeMo Guardrails. Addresses praxis-proxy/ai#48.

The following changes were made:
`providers/nemo.rs`
* Implements NemoProvider which sends a POST to the configured /v1/guardrail/checks endpoint and maps the NeMo response to a GuardResult

`filter.rs`
* on_request_body which checks phase.request (default true), waits for end_of_stream, parses the body, calls extract_messages, invokes the provider, and acts on the verdict:
    > Pass → Continue
    > Block → Reject 403 with the blocked rail name(s) in the body
    > Redact → Continue (body replacement placeholder for praxis-proxy/ai#49)
* extract_messages: parses OpenAI Chat Completions API formats, returns FilterError for unsupported formats

### CODE REVIEW

File: filters/src/guardrails/filter.rs [hunk 1]
```diff
@@ -106,10 +108,71 @@ impl HttpFilter for AiGuardrailsFilter {
...
+/// - OpenAI Chat request: `{"messages": [...]}`
+///
+/// Returns an error for unrecognized body formats to prevent
+/// silently skipping guardrail evaluation.
+///
+/// # Errors
+///
+/// Returns [`FilterError`] if the body is not valid JSON or does not
+/// contain a recognizable messages field.
+fn extract_messages(body: &Bytes) -> Result<Vec<serde_json::Value>, FilterError> {
```
Comment (aslakknutsen) [inline filters/src/guardrails/filter.rs:186]: spec: Common message extraction from OpenAI Chat and MCP body formats

This seems to currently only support the OpenAI Chat format and is missing MCP JSON-RPC

File: filters/src/guardrails/providers/nemo.rs [hunk 2]
```diff
@@ -32,45 +39,141 @@ fn default_timeout_ms() -> u64 {
...
+            messages,
+        };
+        let response = self
+            .client
+            .post(&self.endpoint)
+            .json(&payload)
+            .send()
+            .await
+            .map_err(|e| -> FilterError { format!("ai_guardrails (nemo): failed to send request: {e}").into() })?;
+        if let Some(len) = response.content_length()
```
Comment (aslakknutsen) [inline filters/src/guardrails/providers/nemo.rs]: security: NeMo response without Content-Length. MAX_RESPONSE_SIZE check skipped; response.text() reads unbounded body.

impact: Worker memory exhaustion / DoS

File: tests/integration/tests/suite/examples/guardrails.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    );
+    let proxy = start_proxy(&config);
+
+    let (status, _body) = http_post(
+        proxy.addr(),
+        "/v1/guardrail/checks",
+        r#"{"model":"test","messages":[{"role":"user","content":"hello"}]}"#,
+    );
+
+    assert_ne!(status, 200, "provider down should not forward request to upstream");
```
Comment (aslakknutsen) [inline tests/integration/tests/suite/examples/guardrails.rs]: assert_ne!(status, 200) might allow any number of unrelated return codes to fall through here. 

Something like  assert_eq!(500) instead.

File: filters/Cargo.toml [hunk 4]
```diff
@@ -28,10 +28,12 @@ praxis-ai-apis = { workspace = true }
 praxis-core = { workspace = true }
```
Comment (aslakknutsen) [inline filters/Cargo.toml]: unsued base64 dependency

File: filters/src/guardrails/providers/mod.rs [hunk 5]
```diff
@@ -68,10 +64,6 @@ impl GuardResult {
     /// Returns the status label written to [`FilterResultSet`].
     ///
     /// [`FilterResultSet`]: praxis_filter::FilterResultSet
-    #[cfg_attr(
-        not(test),
-        expect(dead_code, reason = "used once provider evaluation emits results (#578)")
-    )]
     pub fn status_label(&self) -> &'static str {
```
Comment (aslakknutsen) [inline filters/src/guardrails/providers/mod.rs:67]: Doc references FilterResultSet but on_request_body never sets ctx.filter_results which would allow branch-routing.

**Review by aslakknutsen — CHANGES_REQUESTED** [review]

File: filters/src/guardrails/filter.rs (hunk 1 above)
Comment (christinaexyou) [reply filters/src/guardrails/filter.rs:186]: MCP body formats are out-of-scope bc tool-calling isn't fully supported in NeMo Guardrails. I'm currently working on adding tool-calling support upstream and once that's done, I'll add message extraction for MCP body formats

File: filters/Cargo.toml (hunk 4 above)
Comment (christinaexyou) [reply filters/Cargo.toml]: Assuming you mean `praxis-core` instead of `base64`, I removed it

File: filters/src/guardrails/providers/mod.rs (hunk 5 above)
Comment (christinaexyou) [reply filters/src/guardrails/providers/mod.rs:67]: Thanks updated `filters/src/guardrails/filter.rs` so that we pass the NeMo response to the `FilterResultSet`

File: filters/src/guardrails/providers/nemo.rs (hunk 2 above)
Comment (christinaexyou) [reply filters/src/guardrails/providers/nemo.rs]: Done - added a check for response bodies without the Context-Length header

File: filters/Cargo.toml (hunk 4 above)
Comment (aslakknutsen) [reply filters/Cargo.toml]: I meant base64, it was just not in the diff so I gave it a bit of a random line in the file. hehe

Comment (christinaexyou) [reply filters/Cargo.toml]: I ran `RUSTFLAGS="-W unused-crate-dependencies" cargo build -p praxis-ai-filters 2>&1 | grep -i "base64\"`, and the output is empty.  It looks like it's being used [filters/src/agentic/mcp/broker/mod.rs](https://github.com/praxis-proxy/ai/blob/main/filters/src/agentic/mcp/broker/mod.rs#L605)

Comment (christinaexyou) [reply filters/Cargo.toml]: It was introduced in a separate commit that was merged last week haha

File: filters/src/guardrails/filter.rs [hunk 6]
```diff
@@ -105,11 +106,83 @@ impl HttpFilter for AiGuardrailsFilter {
...
+    let json: serde_json::Value = serde_json::from_slice(body)
+        .map_err(|e| -> FilterError { format!("ai_guardrails: request body is not valid JSON: {e}").into() })?;
+
+    // OpenAI Chat format: {"messages": [...]}
+    if let Some(messages) = json.get("messages").and_then(|m| m.as_array()) {
+        return Ok(messages.clone());
```
Comment (aslakknutsen) [inline filters/src/guardrails/filter.rs]: messages.clone() duplicates the full messages Vec<serde_json::Value> after parsing the body. AGENTS.md guidance prefers avoiding clones on request data unless necessary; moving messages out of the parsed Value (e.g. take from object) would avoid an extra allocation on every guarded request.

File: filters/src/guardrails/providers/nemo.rs [hunk 7]
```diff
@@ -32,45 +40,173 @@ fn default_timeout_ms() -> u64 {
...
+    let status = response.status();
+    if !status.is_success() {
+        return Err(format!("ai_guardrails (nemo): provider returned HTTP status code {status}").into());
+    }
+    Ok(())
+}
+
+/// Read the response body, enforcing [`MAX_RESPONSE_SIZE`] as a safety net for
+/// chunked/headerless responses that bypass the `Content-Length` check.
+async fn read_response_body(response: reqwest::Response) -> Result<Bytes, FilterError> {
```
Comment (aslakknutsen) [inline filters/src/guardrails/providers/nemo.rs:164]: Attack surface: compromised or malicious NeMo endpoint (or MITM on the configured endpoint) returning chunked or lying Content-Length responses. Exploit: attacker streams a body far larger than MAX_RESPONSE_SIZE; read_response_body calls response.bytes().await which buffers the entire payload into memory before the len() > MAX_RESPONSE_SIZE check at nemo.rs:169. Impact: worker memory exhaustion / DoS on the proxy worker handling the request. Defense gap: no incremental read with early abort; post-read check only rejects after allocation. Mitigation: stream with a running byte counter (e.g. bytes_stream + take(MAX+1)) and abort before buffering excess data.

File: filters/src/guardrails/providers/nemo.rs [hunk 8]
```diff
@@ -32,45 +40,173 @@ fn default_timeout_ms() -> u64 {
...
+        },
+        other => Err(format!("ai_guardrails (nemo): unknown status '{other}'").into()),
+    }
+}
+
+/// Collect the names of all rails whose `status` is `"blocked"` from the
+/// `rails_status` map and join them with `", "` in sorted order.
+///
+/// Returns an empty string if `rails_status` is absent or no rails are blocked.
+fn blocked_rail_names(rails_status: Option<&serde_json::Value>) -> String {
```
Comment (aslakknutsen) [inline filters/src/guardrails/providers/nemo.rs:204]: blocked_rail_names sorting, empty rails_status, absent map, and non-blocked rail filtering are untested. Strings in 403 responses depend on this logic.

File: filters/src/guardrails/filter.rs [hunk 9]
```diff
@@ -105,11 +106,83 @@ impl HttpFilter for AiGuardrailsFilter {
...
+/// - OpenAI Chat request: `{"messages": [...]}`
+///
+/// Returns an error for unrecognized body formats to prevent
+/// silently skipping guardrail evaluation.
+///
+/// # Errors
+///
+/// Returns [`FilterError`] if the body is not valid JSON or does not
+/// contain a recognizable messages field.
+fn extract_messages(body: &Bytes) -> Result<Vec<serde_json::Value>, FilterError> {
```
Comment (aslakknutsen) [inline filters/src/guardrails/filter.rs:186]: Missing direct tests for invalid JSON, missing messages key, messages not an array, and empty messages array. Fail-closed behavior for unsupported bodies is a spec requirement but only partially exercised end-to-end.

File: filters/src/guardrails/tests.rs [hunk 10]
```diff
@@ -6,6 +6,36 @@ use super::{
     filter::AiGuardrailsFilter,
 };
 
+// =============================================================================
```
Comment (aslakknutsen) [inline filters/src/guardrails/tests.rs:9]: tests.rs now covers pass (on_request_body_passes_through), block (on_request_body_blocked_writes_filter_results), and modified (on_request_body_modified_writes_filter_results) via wiremock, but unit-level coverage still omits: end_of_stream=false early continue, phase.request=false skip, None/empty body continue, invalid request JSON FilterError, missing messages FilterError, NeMo HTTP non-2xx FilterError, NeMo invalid JSON FilterError.

**Review by aslakknutsen — CHANGES_REQUESTED** [review]

File: filters/src/guardrails/providers/nemo.rs (hunk 7 above)
Comment (christinaexyou) [reply filters/src/guardrails/providers/nemo.rs:164]: ack - replaced with incremental checks; for every chunk, we append the response with the chunk and check if it exceeds the max response length

File: filters/src/guardrails/providers/nemo.rs (hunk 8 above)
Comment (christinaexyou) [reply filters/src/guardrails/providers/nemo.rs:204]: thanks for pointing this out - added tests

File: filters/src/guardrails/tests.rs (hunk 10 above)
Comment (christinaexyou) [reply filters/src/guardrails/tests.rs:9]: done - updated the tests to cover those cases

File: filters/src/guardrails/filter.rs (hunk 9 above)
Comment (christinaexyou) [reply filters/src/guardrails/filter.rs:186]: added tests to `filters/src/guardrails/tests.rs` and `tests/integration/tests/suite/examples/guardrails.rs`

**Comment by christinaexyou** [comment]
@aslakknutsen thanks so much for the detailed review !

File: examples/configs/nemo-guardrails.yaml [hunk 11]
```diff
@@ -0,0 +1,52 @@
+# Guardrails (NeMo)
```
Comment (aslakknutsen) [inline examples/configs/nemo-guardrails.yaml:1]: docs/developing/conventions.md requires updating examples/README.md when adding example configs. The new nemo-guardrails.yaml is absent from the General configs table, so discoverability lags the implementation.

**Review by aslakknutsen — CHANGES_REQUESTED** [review]

File: examples/configs/nemo-guardrails.yaml (hunk 11 above)
Comment (christinaexyou) [reply examples/configs/nemo-guardrails.yaml:1]: ack updated `examples/README.md`

**Review by aslakknutsen — APPROVED** [review]

**Comment by aslakknutsen** [comment]
@christinaexyou You need to sign the commits, besides that it looks good :)

**Comment by leseb** [comment]
please use this as a reference to replace callout client https://github.com/praxis-proxy/ai/pull/599

**Comment by christinaexyou** [comment]
@leseb thanks for your help unblocking my PR ! fyi the NeMoProvider never used `CalloutClient` directly and requires [praxis-core#826](https://github.com/praxis-proxy/praxis/issues/826) to replace it with `SubRequest`
